### PR TITLE
PR backend to master

### PR DESCRIPTION
**Pull request by cocolocomoco21**
This PR merges backend branch to master. Included in the PR is:

- Implement Server.java, which is the entry point of the program
- Implement *Controllers.java, which are the controllers for the various resources (e.g. Tutor, Course, Shift, Schedule, etc.)
- Implement data model described in our class diagram
- Initial testing classes skeleton and include JUnit4 framework 
- Update README to be (close to) submission-ready level (for the Iteration 1 review by TAs)
- Make Gradle work with our dependencies (build.gradle -> Sparkjava, Gson, SLF4J, JUnit4) and have the tasks we want (build, run)

This PR should contain nearly everything the FE needs at this time as well, though small additions may be required for the Iteration 1 deadline.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,18 +1,77 @@
...
-To grab dependencies and run the UI, run:
+### Electron UI
+**Dependencies:** TODO. npm
+
+To grab dependencies and build the UI, from the top-level directory, run:
 ``` 
 cd electron-src
-npm install
-npm start
+sudo npm install
```
Comment (collielimabean) [inline README.md]: Nitpick, but sudo is not necessary to install packages locally. If you do, we've got problems...

Comment (cocolocomoco21) [reply README.md]: I agree, it shouldn't be. I didn't say this, but the frontend section will still be modified after/during the frontend merge to master. I was having difficulties installing on Ubuntu when I was attempting to build fresh, but sudo helped (??), so it was left here.

File: src/main/java/ulcrs/models/session/Session.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package ulcrs.models.session;
+
+import ulcrs.models.schedule.Schedule;
+
+public class Session {
+
+    private String name;
+    private Schedule existingSchedule;
+    private long generatedTimestamp;
```
Comment (collielimabean) [inline src/main/java/ulcrs/models/session/Session.java]: Use full DateTime objects or whatever Java's or JodaTime's equivalent is.

Comment (cocolocomoco21) [reply src/main/java/ulcrs/models/session/Session.java]: Ahh yeah good catch, will update.

File: src/main/java/ulcrs/Server.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    private void initializeControllers() {
+        this.tutorController = new TutorController();
+        this.courseController = new CourseController();
+        this.scheduleController = new ScheduleController();
+        this.shiftController = new ShiftController();
+        this.sessionController = new SessionController();
+    }
+
+}
```
Comment (Wrn2x) [inline src/main/java/ulcrs/Server.java]: Ending newline

File: src/main/java/ulcrs/controllers/CourseController.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    private List<Course> getCourseList(Request request, Response response) {
+        response.type(CONTENT_TYPE_JSON);
+
+        InputStream is = getClass().getClassLoader().getResourceAsStream("mockCoursesNoTime.json");
+        JsonReader reader = new JsonReader(new InputStreamReader(is));
+
+        List<Course> courses = gson.fromJson(reader, new TypeToken<List<Course>>() {}.getType());
+        return courses;
```
Comment (Wrn2x) [inline src/main/java/ulcrs/controllers/CourseController.java]: Return all course details or just course identifiers?

File: src/main/resources/mockCoursesFull.json [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                        }
+                    }
+                }
+            ],
+            "requiredShiftAmount": 2,
+            "preferredShiftAmount": 4,
+            "intensity": "HIGH"
+        }
+    }
+]
```
Comment (Wrn2x) [inline src/main/resources/mockCoursesFull.json]: Ending newline

File: src/main/resources/mockCoursesNoTime.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                    "id": 2,
+                    "day": "WEDNESDAY"
+                }
+            ],
+            "requiredShiftAmount": 2,
+            "preferredShiftAmount": 4,
+            "intensity": "HIGH"
+        }
+    }
+]
```
Comment (Wrn2x) [inline src/main/resources/mockCoursesNoTime.json]: Ending newline

File: src/main/java/ulcrs/models/session/Session.java (hunk 2 above)
Comment (cocolocomoco21) [reply src/main/java/ulcrs/models/session/Session.java]: Updated to LocalTime, not a date, as Jason pointed out in lecture. Will update this again.

File: src/main/java/ulcrs/Server.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        Server server = startServer();
+
+        // Route to the proper methods
+        Spark.path("/ulcrs", () -> {
+            // Paths for resources, handled by appropriate Controllers
+            Spark.path("/tutor", server.tutorController.routes());
+            Spark.path("/course", server.courseController.routes());
+            Spark.path("/schedule", server.scheduleController.routes());
+            Spark.path("/shift", server.shiftController.routes());
+            Spark.path("/session", server.sessionController.routes());
```
Comment (cocolocomoco21) [inline src/main/java/ulcrs/Server.java]: For these Spark methods, Spark recommends you use static methods. This isn't vital as it still works, but it's something to consider

File: src/main/java/ulcrs/Server.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+     * Create and start the Server object used to route API calls. The Server object simply acts as a container for
+     * API routing facilitated by Sparkjava and controllers used in this routing.
+     *
+     * @return Server - the Server object to act as the server
+     */
+    private static Server startServer() {
+        Server server = new Server();
+
+        // Must be done before route mapping
+        Spark.staticFiles.location("/");
```
Comment (cocolocomoco21) [inline src/main/java/ulcrs/Server.java]: Ditto from other comment

File: src/main/java/ulcrs/controllers/CourseController.java (hunk 4 above)
Comment (cocolocomoco21) [reply src/main/java/ulcrs/controllers/CourseController.java]: This should only return the data the FE needs to represent courses. That data should be correct in the json file we're using here. Just ID's don't give important things, like name and shifts, that the FE needs from this call.

**Pull request merged**